subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1132 @@
+	struct k_poll_signal read_signal;
+	/** buffer for @a recv_q recv_q */
+	u8_t buf[CONFIG_NET_SOCKETPAIR_BUFFER_SIZE];
+};
+
+/* forward declaration */
+static const struct socket_op_vtable spair_fd_op_vtable;
+
+#undef sock_is_nonblock
+/** Determine if a @ref spair is in non-blocking mode */
+static inline bool sock_is_nonblock(const struct spair *spair)
+{
+	return !!(spair->flags & SPAIR_FLAG_NONBLOCK);
+}
+
+/** Determine if a @ref spair is connected */
+static inline bool sock_is_connected(const struct spair *spair)
+{
+	const struct spair *remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	if (remote == NULL) {
+		return false;
+	}
+
+	return true;
+}
+
+#undef sock_is_eof
+/** Determine if a @ref spair has encountered end-of-file */
+static inline bool sock_is_eof(const struct spair *spair)
+{
+	return !sock_is_connected(spair);
+}
+
+/**
+ * Determine bytes available to write
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking.
+ */
+static inline size_t spair_write_avail(struct spair *spair)
+{
+	struct spair *const remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	struct k_pipe *const p = (remote == NULL)
+		? NULL
+		: &remote->recv_q;
+
+	if (remote == NULL) {
+		return 0;
+	}
+
+	return k_pipe_write_avail(p);
+}
+
+/**
+ * Determine bytes available to read
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * read from a given @ref spair without blocking.
+ */
+static inline size_t spair_read_avail(struct spair *spair)
+{
+	struct k_pipe *const p = &spair->recv_q;
+
+	return k_pipe_read_avail(p);
+}
+
+/** Swap two 32-bit integers */
+static inline void swap32(u32_t *a, u32_t *b)
+{
+	u32_t c;
+
+	c = *b;
+	*b = *a;
+	*a = c;
+}
+
+/**
+ * Delete @param spair
+ *
+ * This function deletes one endpoint of a socketpair.
+ *
+ * Theory of operation:
+ * - we have a socketpair with two endpoints: A and B
+ * - we have two threads: T1 and T2
+ * - T1 operates on endpoint A
+ * - T2 operates on endpoint B
+ *
+ * There are two possible cases where a blocking operation must be notified
+ * when one endpoint is closed:
+ * -# T1 is blocked reading from A and T2 closes B
+ *    T1 waits on A's write signal. T2 triggers the remote
+ *    @ref spair.write_signal
+ * -# T1 is blocked writing to A and T2 closes B
+ *    T1 is waits on B's read signal. T2 triggers the local
+ *    @ref spair.read_signal.
+ *
+ * If the remote endpoint is already closed, the former operation does not
+ * take place. Otherwise, the @ref spair.remote of the local endpoint is
+ * set to -1.
+ *
+ * If no threads are blocking on A, then the signals have no effect.
+ *
+ * The memeory associated with the local endpoint is cleared and freed.
+ */
+static void spair_delete(struct spair *spair)
+{
+	int res;
+
+	struct spair *remote = NULL;
+	bool have_remote_sem = false;
```

Comment:
Unnecessary empty line here, also looks better if you just do 
```
	struct spair *remote = NULL;
	bool have_remote_sem = false;
	int res;
```

Reply:
I always like to put the return variable at the top, and also uninitialized variables before initialized variables.

Reply:
I'll remove the empty line. Typically I like to put uninitialized variables before initialized variables though